### ISSUE TITLE
Debug: Provide graceful shutdown on debug stop action

### ISSUE THREAD

**Issue by circlesmiler** [microsoft/vscode-java-debug#1274]
When stopping a debugging session, the application will be killed immediately without running `shutdownHook` code. There is the same discussion in the GO Lang project (https://github.com/golang/vscode-go/issues/120).

It would be nice to have the chance to decide, how the application will be killed (`SIGKILL` vs `SIGTERM`). Both are valid use-cases. Maybe it's even possible to have options on the STOP button of the debug panel. As we have multi-project workspaces, it really depends on the application (even situation) what you need.

##### Environment
- Operating System: MacOS (But same on Windows/Linux)
- JDK version: 11
- Visual Studio Code version: 1.74.0
- Java extension version: v0.25.7
- Java Debugger extension version: v0.47.0

##### Steps To Reproduce
1. Made a java application with an `shutdownHook` ([Runtime.addShutdownHook](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html))
2. Start the application
3. Stop the application by clicking the STOP button in the debug panel.

##### Current Result
The `shutdownHook` will not be executed.

##### Expected Result
The `shutdownHook` will be executed. (At least if the user chooses to)

##### Additional Informations
N/A

**Comment by testforstephen** [comment]
Since the debugger is using JDI [VirtualMachine#exit(int)](https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jdi/com/sun/jdi/VirtualMachine.html#exit(int)) API to stop the target debuggee, the debugger has little control over this behavior. I'm wondering if other IDEs support the capability you declare?

**Comment by circlesmiler** [comment]
@testforstephen IntelliJ does. 🤷‍♂️ We noticed because our Spring Boot project does some clean up stuff, when shutting down. The IntelliJ users have no problem, but VSCode users have. I'm not sure how IntelliJ is doing it.

**Comment by stefanrybacki** [comment]
@testforstephen Eclipse also shuts down application gracefully

**Comment by robaho** [comment]
The debug panel should work like the Node.js debug extension. The first press of STOP uses SIGINT. The second press of stop sends SIGKILL.

The 'restart' command in the debug panel should work similarly - or if the SIGINT doesn't kill the process in X seconds (configurable in the launch.json) a SIGKILL is sent. Then when the process dies and new one is created.

This is a serious problem because not running shutdown hooks can leave many applications in a corrupted state.

**Comment by robaho** [comment]
I believe the offending code is [here](https://github.com/microsoft/java-debug/blob/main/com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugSession.java#L123) - which is in the 

This should first attempt to get the Process object, and if non-null, use process.destroy() else use the vm.exit()

**Comment by totszwai** [comment]
Yeah, I second this, I couldn't test my graceful shutdown procedure through VS2022 debugging.
Namely, my process starts other subprocesses and during its shutdown, is supposed to kill the child processes, but right now the parent just dies immediately leaving the child hanging and need to be killed via task manager.

**Comment by mozhuanzuojing** [comment]
现在有消息吗？

**Comment by jrx-sjg** [comment]
How is this issue closed when the forced shutdown is still hapening?
